File: lib-cli/tmp/server/src/producer/implementation/handlers/users_request.rs

use super::{
    identification, pack, producer::Control, protocol, responses, Context, HandlerError,
    ProducerError,
};
use fiber::server;

pub async fn process<E: std::error::Error, C: server::Control<E> + Send + Clone>(
    identification: &mut identification::Identification,
    filter: &identification::Filter,
    context: &mut Context,
    request: &protocol::Users::Request,
    sequence: u32,
    control: &Control<E, C>,
) -> Result<(), HandlerError> {
    let uuid = identification.uuid();
    let buffer = match responses::users_request::response(
        identification,
        filter,
        context,
        request,
        control,
    )
    .await
    {
        Ok(mut response) => pack(&sequence, &uuid, &mut response)?,
        Err(mut error) => pack(&sequence, &uuid, &mut error)?,
    };
    control
        .send(buffer, Some(uuid))
        .await
        .map_err(|e: ProducerError<E>| HandlerError::Processing(e.to_string()))
}

File: lib-cli/tmp/server-ts/src/producer/implementation/responses/users.request.ts
import { Producer, Identification, Filter, Context, Protocol } from "./index";
import { response } from "../../responses/users.request";

export class Response {
	private _response!: Protocol.Users.Response | Protocol.Users.Err;

	constructor(res: Protocol.Users.Response | Protocol.Users.Err) {
		this._response = res;
	}

	public pack(sequence: number, uuid: string): ArrayBufferLike {
		return this._response.pack(sequence, uuid);
	}
}

export function handler(
	request: Protocol.Users.Request,
	consumer: Identification,
	filter: Filter,
	context: Context,
	producer: Producer,
	sequence: number
): Promise<void> {
	return response(request, consumer, filter, context, producer).then(
		(res) => {
			return producer.send(
				consumer.uuid(),
				res.pack(sequence, consumer.uuid())
			);
		}
	);
}
